### PR TITLE
extmod/modlwip: socket polling behaviour fixes

### LINKED ISSUE #3449: Polling non-blocking socket for POLLOUT after receiving RST results in null pointer dereference
Please run the attached script in a esp8266 board with modifying the target address/port to a reachable host and closed port.

[poll.py](https://github.com/micropython/micropython/files/1490567/trys.py.txt)

### PR DESCRIPTION
**Pull request by rkojedzinszky**
This adds support for returning POLLHUP and POLLERR flags. Also, for
listening sockets, also returns POLLIN when a client is connected.

Fixes #3449

### CODE REVIEW

File: extmod/modlwip.c [hunk 1]
```diff
@@ -1168,16 +1170,17 @@ STATIC mp_uint_t lwip_socket_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr_
             ret |= MP_STREAM_POLL_RD;
         }
 
-        if (flags & MP_STREAM_POLL_WR && tcp_sndbuf(socket->pcb.tcp) > 0) {
+        if (flags & MP_STREAM_POLL_WR && (socket->pcb.tcp == NULL || tcp_sndbuf(socket->pcb.tcp) > 0)) {
```
Comment (dpgeorge) [inline extmod/modlwip.c]: Shouldn't this be `(socket->pcb.tcp != NULL && ...`?  Is it still writable if the pcb is NULL?

File: extmod/modlwip.c [hunk 2]
```diff
@@ -1168,16 +1170,17 @@ STATIC mp_uint_t lwip_socket_ioctl(mp_obj_t self_in, mp_uint_t request, uintptr_
...
             ret |= MP_STREAM_POLL_WR;
         }
 
-        if (socket->state == STATE_PEER_CLOSED) {
-            // Peer-closed socket is both readable and writable: read will
-            // return EOF, write - error. Without this poll will hang on a
-            // socket which was closed by peer.
-            ret |= flags & (MP_STREAM_POLL_RD | MP_STREAM_POLL_WR);
-        }
```
Comment (dpgeorge) [inline extmod/modlwip.c]: Why is this part removed, how is this "peer closed" state handled now?

File: extmod/modlwip.c [hunk 3]
```diff
@@ -720,6 +720,8 @@ STATIC mp_obj_t lwip_socket_listen(mp_obj_t self_in, mp_obj_t backlog_in) {
     socket->pcb.tcp = new_pcb;
     tcp_accept(new_pcb, _lwip_tcp_accept);
 
+    socket->state = STATE_CONNECTING;
```
Comment (dpgeorge) [inline extmod/modlwip.c]: Arguable, a listening socket is still "new" and not yet connecting because there's not necessarily anyone trying to connect to it.  What problem does this line solve?  Is there a test case that can be written which shows why this line should be here?

**Review by dpgeorge — COMMENTED** [review]
Thanks for the updated PR.  Because socket polling behaviour is subtle and critical to get right, any changes to functionality really need to be accompanied by tests which show that the old behaviour is/was wrong and that the patch fixes the behaviour.

For the case of a closed socket leading to a seg-fault, that's clear what is wrong and how to fix it, and tests are easy to write.  But for the other changes it's not clear (to me at least) how and why things are changed.

File: extmod/modlwip.c (hunk 3 above)
Comment (rkojedzinszky) [reply extmod/modlwip.c]: A listening socket will never be in other state other than new, but it should not return HUP or ERR anyway. Another condition will set HUP if the connection is not CONNECTING nor CONNECTED either.

Please find my simple test script here: https://gist.github.com/rkojedzinszky/cbb6710c0cd6b205eda6f32b4bfcd7fc

File: extmod/modlwip.c (hunk 1 above)
Comment (rkojedzinszky) [reply extmod/modlwip.c]: On posix systems, when a socket is closed, it is being writable, where a write operation will return an error.

File: extmod/modlwip.c (hunk 2 above)
Comment (rkojedzinszky) [reply extmod/modlwip.c]: The closed socket is handled with HUP/ERR flag.

Again check this: https://gist.github.com/rkojedzinszky/cbb6710c0cd6b205eda6f32b4bfcd7fc

**Comment by dpgeorge** [comment]
Sorry for taking so long to get back to this, and thanks for the detailed test script and comments.  I ran your test (polltry.py) on the latest version with the patch 318f874cda22567a55bb004434ed0ec891fd7d61 and that has fixed a few cases.  The remaining case that is fixed by this PR here is how to handle a socket in STATE_NEW, namely to set POLLHUP.  In 1e2a6a84a263e069d7be5f27d73e7fc124413fa5 I incorporated this behaviour into the existing logic, which can be seen to be a straightforward change that shouldn't impact any existing behaviour because it just sets an additional flag (POLLHUP) for a case that was not considered before (STATE_NEW).

The polltry.py test running on esp8266 still doesn't behave exactly the same as CPython on Linux, but it's closer.  One of the things to improve is that poll should set POLLHUP/POLLERR regardless of the requested flags.  But changing this behaviour will change how select.select works so needs to be done separately.
